WordPress/gutenberg
# List View: Allow dragging outside the immediate area by passing down a dropZoneElement

## What?


Part of addressing #43881 and #32438, alternative to #49420 and #50689. Overall, this PR makes dragging to the top and bottom positions of the list view easier.

Allow dragging and dropping within the list view, beyond the bounds of the list view itself, when passed a `dropZoneElement` (an element to be used as the drop zone, typically a parent of the final drop zone ref). Within the list view sidebar, pass the list view container area as a `dropZoneElement` to the `ListView` component.

Then, when the mouse leaves that drop zone area, clear the drop target so that the blue line no longer appears. The result should be that the blue drop indicator line within the list view is now accurate — it will only display when a drop will occur, and it displays across the whole list view area, making it easier to drag to the top and bottom-most positions.

## Why?


One of the frustrations with drag and drop in the list view is that it's very difficult to drag to the top and bottom-most positions. This PR hopefully makes that experience feel more flexible and accurate.

## How?


* Pass an optional `dropZoneElement` element to `useDropZone` and use its element if available, falling back to the one passed down by `useDropZone`.
* Allow this prop at the `ListView` and `useListViewDropZone` levels.
* In the post and site editors (and widget editor), pass a ref of the list view container to the `ListView`.
* Within the list view drop zone logic, when the user leaves the drop zone area, set the target to `null` so that the drop indicator disappears.
* Update nesting logic so that if a user drags below the final item in the list, it won't be treated as a nesting gesture. This helps when a collapsed block is the last item in the list view, as a user dragging all the way to the bottom-most position likely wishes to move the block below rather than nest.

## Testing Instructions





1. In the post editor and site editors, experiment with dragging blocks to the top-most and bottom-most positions. Compared with `trunk`, it should be a lot easier to drop to these positions.
2. Test with short list views and really long post content.
3. If the last block in the list is a collapsed container block (e.g. group), then the nesting gesture should only apply when dragging over the bottom half of that block — if you drag further down, the drop indicator should change to be beneath that block rather than inside it.
4. If the last block in the list is an expanded container block with some level of nesting, then dragging to parents should still work when dragging well below the bottom block position.
5. Smoke test that dragging regular blocks in the editor canvas works as on trunk.
6. Smoke test other list view instances (e.g. navigation in editor browse mode) that nothing breaks elsewhere.

## Screenshots or screencast 

Note how in the before version, the Heading looks like it can be dropped to the very bottom of the area by dragging well below, but it cannot. In the after version, the drop is respected correctly.

| Before | After |
| --- | --- |
| ![2023-05-18 17 17 20](https://github.com/WordPress/gutenberg/assets/14988353/766f276b-2cd7-4e66-9b16-2d6ac9ff769d) | ![2023-05-18 17 13 38](https://github.com/WordPress/gutenberg/assets/14988353/6e1d58e4-af37-414a-ad7d-f762cda2fb54) |

Some more details of the behaviour in this PR:

| Dragging up the block hierarchy should still work when cursor is lower than the last block | Nesting a collapsed block only happens if the cursor is over the block (makes it easier to drop below) |
| --- | --- |
| ![2023-05-19 12 00 02](https://github.com/WordPress/gutenberg/assets/14988353/abd06731-3cbb-4321-9809-f37a591022ab) | ![2023-05-19 11 59 31](https://github.com/WordPress/gutenberg/assets/14988353/c390335e-504c-431d-b9a7-a7db0b93e166) |
